operator/controllers/contour_ingress.go

```diff
@@ -0,0 +1,291 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	contour "github.com/projectcontour/contour/apis/projectcontour/v1"
+	machinelearningv1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
+	v1 "github.com/seldonio/seldon-core/operator/apis/machinelearning.seldon.io/v1"
+	"github.com/seldonio/seldon-core/operator/constants"
+	v13 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	types2 "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"knative.dev/pkg/kmp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	ENV_CONTOUR_ENABLED   = "CONTOUR_ENABLED"
+	ENV_CONTOUR_BASE_HOST = "CONTOUR_BASE_HOST"
+)
+
+var grpcProtocol = "h2c"
+
+// +kubebuilder:rbac:groups=projectcontour.io,resources=httpproxies,verbs=get;list;watch;create;update;patch;delete
+
+type ContourIngress struct {
+	client   client.Client
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+func NewContourIngress() Ingress {
+	return &ContourIngress{}
+}
+
+func (i *ContourIngress) AddToScheme(scheme *runtime.Scheme) {
+	contour.AddKnownTypes(scheme)
+}
+
+func (i *ContourIngress) SetupWithManager(mgr ctrl.Manager) ([]runtime.Object, error) {
+	// Store the client, recorder and scheme for use later
+	i.client = mgr.GetClient()
+	i.recorder = mgr.GetEventRecorderFor(constants.ControllerName)
+	i.scheme = mgr.GetScheme()
+
+	// Index on VirtualService
+	if err := mgr.GetFieldIndexer().IndexField(&contour.HTTPProxy{}, ownerKey, func(rawObj runtime.Object) []string {
+		// grab the deployment object, extract the owner...
+		vsvc := rawObj.(*contour.HTTPProxy)
+		owner := metav1.GetControllerOf(vsvc)
+		if owner == nil {
+			return nil
+		}
+		// ...make sure it's a SeldonDeployment...
+		if owner.APIVersion != apiGVStr || owner.Kind != "SeldonDeployment" {
+			return nil
+		}
+
+		// ...and if so, return it
+		return []string{owner.Name}
+	}); err != nil {
+		return nil, err
+	}
+	return []runtime.Object{&contour.HTTPProxy{}}, nil
+}
+
+func (i *ContourIngress) GeneratePredictorResources(mlDep *v1.SeldonDeployment, seldonId string, namespace string, ports []httpGrpcPorts, httpAllowed bool, grpcAllowed bool) (map[string][]runtime.Object, error) {
+	contourBaseHost := GetEnv(ENV_CONTOUR_BASE_HOST, "") // TODO(jpg): This to be better to handle cases when base host isn't set better
+	var httpProxies []runtime.Object
+	var routes []contour.Route
+	var grpcRoutes []contour.Route
+
+	for i, predictor := range mlDep.Spec.Predictors {
+		predictorServiceName := v1.GetPredictorKey(mlDep, &predictor)
+		routes = append(routes, contour.Route{
```

Comment:
I notice this is using contour.Route, which I presume corresponds to contour's [IngressRoute concept](https://projectcontour.io/docs/v1.5.0/ingressroute/). I notice contour [also supports plain-old Ingress](https://projectcontour.io/docs/v1.5.0/deploy-options/). Since you're also looking to address https://github.com/SeldonIO/seldon-core/issues/1778, would it make sense to just use Ingress for Contour as well? Or does that miss out on some contour functionality or add some other complication?

Reply:
Actually `contour.Route` doesn't map to `IngressRoute` (that would be `contour.IngressRoute`), it's merely part of the spec of the `contour.HTTPProxy` interface that allows for routing traffic of a particular virtual host by a set of conditions.

Unfortunately the generic Kubernetes`Ingress` interface lacks the required features to implement the behaviours supported in Seldon. If this was a viable option it would already be used for Istio and Ambassador support, both of which are also capable of acting as `Ingress` controller implementations. The main shortcomings are lack of weight based routing, non-path based conditions and mirroring/shadow support. It's unlikely that `Ingress` will evolve to incorporate these more complex routing primitives in the near future but if it does eventually come to fruition that would be ideal.

Reply:
Yeah I just [read a guide on contour](https://octetz.com/docs/2019/2019-04-12-contour-ingress-with-envoy/) and see what you mean. I was wondering if contour might support weighted routing [through annotations like nginx does](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#canary). It is a shame that fudging this stuff in through annotations is all that Ingress supports and understandable that many projects are choosing to define their own CRDs instead. Makes it awkward for us. I guess you're going to prioritise Contour support specifically?